Prevent blank revisions from being read from the backend

rosstimothy opened the pull request:
Overwrites any empty resource revisions with a placeholder value to prevent any blank revisions from being provided to users.

rosstimothy committed:
Prevent blank revisions from being read from the backend

Overwrites any empty resource revisions with a placeholder value to
prevent any blank revisions from being provided to users.

fspmarshall commented on lib/backend/backend.go:
```diff
@@ -433,6 +433,12 @@ func CreateRevision() string {
+// the revision of the item in the backend is empty. This can happen
+// to any existing resources that were last written before support for
+// revisions was added. The value is the base32 encoding of deadbeef.
+const BlankRevision = "GB4GIZLBMRRGKZLG"
```
If there is a bug, it will be non-obvious that this is a sentinel value.  I'd suggest going with something that is more obviously not a real revision.  Ideally something that contains at least one character that ins't allowable in any real revision encoding (e.g. `@empty`).

espadolini replied:
The only backends where this is relevant also use UUIDs for revisions (i.e. they use `backend.CreateRevision`). Maybe we could use `00000000-0000-0000-0000-000000000000`? It's never generated as a uuid v4, so it's actually impossible (as opposed to practically impossible) for it to be a generated revision of an item after the introduction of revisions.

espadolini reviewed (commented):
Why do we need the blank revision handling for the memory backend?

espadolini commented on lib/backend/dynamo/dynamodbbk.go:
```diff
@@ -572,6 +578,10 @@ func (b *Backend) Delete(ctx context.Context, key []byte) error {
 // ConditionalUpdate updates the matching item in Dynamo if the provided revision matches
 // the revision of the item in Dynamo.
 func (b *Backend) ConditionalUpdate(ctx context.Context, item backend.Item) (*backend.Lease, error) {
+	if item.Revision == backend.BlankRevision {
+		item.Revision = ""
+	}
+
```
If the passed-in revision is the empty string we should fail the operation (I don't know if it should be a `CompareFailed` or a `BadParameter`, but `pgbk` will return `CompareFailed` on anything that's not a UUID, including the blank string).

espadolini commented:
> We either have to insert the place holder into the event stream so that resources in the cache can never have an empty string or add blank revision handling to the memory backend.

It's the job of the backend driver to never produce an empty revision string, if that means changing the event stream code to make it so that the empty revision string never appears even when "it should" then that's what we should do, IMO.

fspmarshall reviewed (approved):
R.E. the memory backend handling blank revisions; I think it should handle them correctly, but if it observes them while in mirror mode it should emit a warning.

espadolini reviewed (approved):
LGTM other than the memory backend

rosstimothy committed:
Prevent backend items with empty revisions

All resource revisions are now being set on a backend.Item before
persisting the item to the backend.

espadolini commented:
Considering the situation with the cache `applyFn`s I think we should reconsider the special handling of the blank revision string in the memory backend in mirror mode: an item of a cached type that's stored in dynamodbbk or litebk that has a blank revision because it's pre-migration will be yielded from the cache with its correct revision until it's overwritten, resulting in very confusing behavior if it's conditionally updated.

espadolini commented on lib/backend/memory/memory.go:
```diff
@@ -386,9 +386,10 @@ func (m *Memory) CompareAndSwap(ctx context.Context, expected backend.Item, repl
 func (m *Memory) ConditionalDelete(ctx context.Context, key []byte, rev string) error {
-	if len(key) == 0 {
-		return trace.BadParameter("missing parameter key")
+	if len(key) == 0 || rev == "" {
```
```suggestion
	if len(key) == 0 || (rev == "" && !m.Mirror) {
```
